File: algo/42.py
from typing import List

# class Solution:
#     def trap(self, height: List[int]) -> int:
#         left, right = 0, len(height) - 1
#         left_max, right_max = 0, 0
#         volumn = 0

#         while left <= right:
#             left_max = max(left_max, height[left])
#             right_max = max(right_max, height[right])

#             if left_max <= right_max:
#                 volumn += left_max - height[left]
#                 left += 1
#             else:
#                 volumn += right_max - height[right]
#                 right -= 1
#         return volumn

class Solution:
    def trap(self, height: List[int]) -> int:
        stack = []
        volumn = 0

        for i in range(len(height)):
            while stack and height[i] > height[stack[-1]]:
                top = stack.pop()
                if not stack:
                    break

                distance = i - stack[-1] - 1
                gap = min(height[i], height[stack[-1]]) - height[top]
                volumn += distance * gap

            stack.append(i)
            
        return volumn

height = [0,1,0,2,1,0,1,3,2,1,2,1]
a = Solution()

print(a.trap(height))

height = [4,2,0,3,2,5]
print(a.trap(height))

File: old/algo/42.kt
// class Solution {
//     fun trap(heights: IntArray): Int {
//         var left = 0
//         var right = heights.size - 1

//         var maxLeftHeight = Int.MIN_VALUE
//         var maxRightHeight = Int.MIN_VALUE
//         var volume = 0

//         while (left < right) {
//             maxLeftHeight = max(heights[left], maxLeftHeight)
//             maxRightHeight = max(heights[right], maxRightHeight)

//             if (maxLeftHeight < maxRightHeight) {
//                 volume += maxLeftHeight - heights[left]
//                 left++
//             } else {
//                 volume += maxRightHeight - heights[right]
//                 right--
//             }
//         }
//         return volume
//     }
// }

// val heights = intArrayOf(0,1,0,2,1,0,1,3,2,1,2,1)
// Solution().trap(heights)

class Solution {
    fun trap(heights: IntArray): Int {
        var volume = 0
        var stack = mutableListOf<Int>()

        for ((i, height) in heights.withIndex()) {
            while (stack.isNotEmpty() && heights[stack[stack.lastIndex]] < height) {
                val bottom = stack.removeAt(stack.lastIndex)
                if (stack.isEmpty()) break
                val lastTop = stack[stack.lastIndex]
                val minTopHeight = min(heights[lastTop], height)
                val distance = i - lastTop - 1
                val diff = minTopHeight - heights[bottom]
                volume += diff * distance
            }
            stack.add(i)
        }

        return volume
    }
}

val heights = intArrayOf(0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1)
Solution().trap(heights)
